
###
###   Lemur v2.0
###   Books to Prisoners Inventory System
###
###   Tom Wiltzius, tom.wiltzius%gmail.com
###   Oct 2010-2016
###

Python dependencies listed in requirements.txt

Also requires a Google Books API key, yarn, webpack, MySQL or sqlite

To run tests:
`./manage.py test`


To run the server locally:
-) sudo apt install -y default-libmysqlclient-dev pkg-config npm
-) sudo npm install --global yarn
-) install client deps with `yarn` in root directory
-) install webpack globally if it isn't with `yarn global add webpack` (webpack 2.5.1 works)
-) Build/watch webpack assets with `./node_modules/webpack/bin/webpack.js` in base project directory
-) pip install -r requirements.txt
-) set up settings_secret.py; use SqlLite database
-) Run development server with `./manage.py runserver 0.0.0.0:8000`

Note: if you want to access your local dev server from another computer, add
`ALLOWED_HOSTS = ['192.168.1.37']` to LemurAptana/settings.py (swapping in the
server's IP address or FQDN in place of the example IP address)

To manage the server in production:
-) Various packages need be installed, including: `apt-get install libffi-dev`
-) Use lemur-fcgi.conf in /etc/init/
-) Query service status, restart, etc with `sudo initctl status lemur-fcgi`
-) Rebuild client, run migrations, and restart service with `./production_restart.sh`

For site administration (production or local)
-) Django admin site is located at `/admin`

Machine dependencies:
-) yarn
-) webpack
-) MySQL or sqlite
-) Ability to build python3 lxml from source or install system package (e.g. ubuntu python3-lxml)
-) libssl-dev on Linux

# Limitations
Unable to install python dependencies on Debian 11 (10 and 12 are untested). Issue is with compiling
the mysqlclient library. This is likely due to the old version of the python library not working with
the new version of default-libmysqlclient-dev.

```
   _mysql.c: In function ‘_mysql_ConnectionObject_ping’:
    _mysql.c:1911:41: error: ‘MYSQL’ {aka ‘struct st_mysql’} has no member named ‘reconnect’
     1911 |  if ( reconnect != -1 ) self->connection.reconnect = reconnect;
          |                                         ^
```

The most likely immediate path forward here is to try this again with Debian 10, which is still
getting security updates.

The long term solution is going to be to update the version of the mysql python library, which
will likely require updating to a newer version of Django, which will likely cascade into many
other libraries.
